pkg/registryserver/registryserver.go

```diff
@@ -74,28 +85,41 @@ func (rs *RegistryServer) RegisterController(ctx context.Context, req *registry.
 	}
 	glog.V(3).Infof("Registering node: %s, endpoint: %s", req.NodeId, req.Endpoint)
 
-	rs.nodeClients[req.NodeId] = &NodeInfo{
+	node := &NodeInfo{
 		NodeID:   req.NodeId,
 		Endpoint: req.Endpoint,
 	}
+	rs.mutex.Lock()
+	rs.nodeClients[req.NodeId] = node
+	rs.mutex.Unlock()
+
+	for l := range rs.listeners {
+		l.OnNodeAdded(ctx, node)
+	}
 
 	return &registry.RegisterControllerReply{}, nil
 }
 
 func (rs *RegistryServer) UnregisterController(ctx context.Context, req *registry.UnregisterControllerRequest) (*registry.UnregisterControllerReply, error) {
-	rs.mutex.Lock()
-	defer rs.mutex.Unlock()
-
 	if req.GetNodeId() == "" {
 		return nil, status.Error(codes.InvalidArgument, "Missing NodeId parameter")
 	}
 
-	if _, ok := rs.nodeClients[req.NodeId]; !ok {
+	rs.mutex.Lock()
+	node, ok := rs.nodeClients[req.NodeId]
+	rs.mutex.Unlock()
+	if !ok {
 		return nil, status.Errorf(codes.NotFound, "No entry with id '%s' found in registry", req.NodeId)
 	}
 
+	for l := range rs.listeners {
```

Comment:
`UnregisterController` is now no longer atomic: if you had two competing `UnregisterController` calls for the same node, both might get pass the "ok" check and then the rest of the code gets executed twice whereas before the second call would have gotten an error.

It's a theoretic problem, I'm just pointing out the drawback of locking just subsections of the function.

Reply:
Another race is a RegisterController that comes in after UnregisterController: if the RegisterController takes the mutex in the middle of UnregisterController, then the controller gets removed by the rest of UnregisterController.

Reply:
@pohly  I addressed these issues https://github.com/avalluri/pmem-CSI/commit/dacc98491ffcf6fad66cc886a4a360e51c645f66 . Now the registry calls are idempotent.